### PR TITLE
Fix weighted average double round up

### PR DESCRIPTION
**Pull request by MathisGD**
Fixes [this](https://github.com/morpho-dao/morpho-utils/pull/41#discussion_r946620510) issue found by @QGarchery.
Unexpected, but it is saving gas !
Notice that the reverts for overflows are much more conservative with this solution, but I don't think that this is much of an issue.

### CODE REVIEW

File: test/TestPercentageMath.sol [hunk 1]
```diff
@@ -201,6 +200,17 @@ contract TestPercentageMath is Test {
         PercentageMath.weightedAvg(x, y, percentage);
     }
 
+    function testWeightedAvgOutOfBounds() public {
```
Comment (MerlinEgalite) [inline test/TestPercentageMath.sol:203]: cannot be fuzzed?

File: test/TestPercentageMath.sol [hunk 2]
```diff
@@ -164,12 +167,9 @@ contract TestPercentageMath is Test {
...
     ) public {
-        vm.assume(percentage <= PERCENTAGE_FACTOR);
-        // prettier-ignore
-        vm.assume(
-            (percentage == 0 || y <= MAX_UINT256_MINUS_HALF_PERCENTAGE / percentage) &&
-            (PERCENTAGE_FACTOR - percentage == 0 || x <= MAX_UINT256_MINUS_HALF_PERCENTAGE / (PERCENTAGE_FACTOR - percentage))
-        );
+        vm.assume(percentage <= 1e4);
+        vm.assume(percentage == 0 || y <= (type(uint256).max - 0.5e4) / percentage);
+        vm.assume(1e4 - percentage == 0 || x <= (type(uint256).max - y * percentage - 0.5e4) / (1e4 - percentage));
```
Comment (MerlinEgalite) [inline test/TestPercentageMath.sol:172]: at the opposite of what is done elsewhere

**Review by MerlinEgalite — APPROVED** [review]
Some comments and questions but overall LGTM

File: test/TestPercentageMath.sol (hunk 2 above)
Comment (MathisGD) [reply test/TestPercentageMath.sol:172]: you mean the literal values instead of using the constants ?

File: test/TestPercentageMath.sol (hunk 1 above)
Comment (MathisGD) [reply test/TestPercentageMath.sol:203]: we could yes

**Review by MerlinEgalite — APPROVED** [review]
LGTM

File: src/math/PercentageMath.sol [hunk 3]
```diff
@@ -112,29 +112,23 @@ library PercentageMath {
...
+        //     <=> percentage > 0 and y > (type(uint256).max - percentage) / percentage
         // or if
-        //     x2 = y * percentage + HALF_PERCENTAGE_FACTOR > type(uint256).max
-        //     <=> y * percentage > type(uint256).max - HALF_PERCENTAGE_FACTOR
-        //     <=> percentage > 0 and y > (type(uint256).max - HALF_PERCENTAGE_FACTOR) / percentage
-        //
-        // Note: x1 / PERCENTAGE_FACTOR + x2 / PERCENTAGE_FACTOR <= x * (1 - p) + 1 + y * p + 1 where p is the fractional number percentage / PERCENTAGE_FACTOR
-        // <= max(x, y) + 2 <= type(uint256).max - HALF_PERCENTAGE_FACTOR + 2 <= type(uint256).max
+        //     x * (PERCENTAGE_FACTOR - percentage) + y * percentage + HALF_PERCENTAGE_FACTOR > type(uint256).max
+        //     <=> x > type(uint256).max - y * percentage - HALF_PERCENTAGE_FACTOR / (PERCENTAGE_FACTOR - percentage)
```
Comment (QGarchery) [inline src/math/PercentageMath.sol:119]:
```suggestion
        //     <=> PERCENTAGE_FACTOR - percentage > 0 and  x > (type(uint256).max - y * percentage - HALF_PERCENTAGE_FACTOR) / (PERCENTAGE_FACTOR - percentage)
```

File: src/math/PercentageMath.sol [hunk 4]
```diff
@@ -112,29 +112,23 @@ library PercentageMath {
...
         assembly {
             z := sub(PERCENTAGE_FACTOR, percentage) // Temporary assignment to save gas.
-
             if or(
-                or(gt(percentage, PERCENTAGE_FACTOR), mul(z, gt(x, div(MAX_UINT256_MINUS_HALF_PERCENTAGE, z)))),
-                mul(percentage, gt(y, div(MAX_UINT256_MINUS_HALF_PERCENTAGE, percentage)))
+                gt(percentage, PERCENTAGE_FACTOR),
+                or(
+                    mul(percentage, gt(y, div(MAX_UINT256_MINUS_HALF_PERCENTAGE, percentage))),
+                    mul(z, gt(x, div(sub(sub(MAX_UINT256, mul(y, percentage)), HALF_PERCENTAGE_FACTOR), z)))
```
Comment (QGarchery) [inline src/math/PercentageMath.sol:126]: We can swap the order of sub to have a substraction of a constant by a constant. It should save some gas

File: src/math/PercentageMath.sol [hunk 5]
```diff
@@ -112,29 +112,23 @@ library PercentageMath {
         // Must revert if
         //     percentage > PERCENTAGE_FACTOR
         // or if
-        //     x1 = x * (PERCENTAGE_FACTOR - percentage) + HALF_PERCENTAGE_FACTOR > type(uint256).max
-        //     <=> (PERCENTAGE_FACTOR - percentage) > 0 and x > (type(uint256).max - HALF_PERCENTAGE_FACTOR) / (PERCENTAGE_FACTOR - percentage)
+        //     y * percentage + HALF_PERCENTAGE_FACTOR > type(uint256).max
+        //     <=> percentage > 0 and y > (type(uint256).max - percentage) / percentage
```
Comment (QGarchery) [inline src/math/PercentageMath.sol:116]:
```suggestion
        //     <=> percentage > 0 and y > (type(uint256).max - HALF_PERCENTAGE_FACTOR) / percentage
```

File: test/TestPercentageMath.sol [hunk 6]
```diff
@@ -165,10 +168,10 @@ contract TestPercentageMath is Test {
         uint16 percentage
     ) public {
         vm.assume(percentage <= PERCENTAGE_FACTOR);
-        // prettier-ignore
+        vm.assume(percentage == 0 || y <= (MAX_UINT256 - HALF_PERCENTAGE_FACTOR) / percentage);
         vm.assume(
-            (percentage == 0 || y <= MAX_UINT256_MINUS_HALF_PERCENTAGE / percentage) &&
-            (PERCENTAGE_FACTOR - percentage == 0 || x <= MAX_UINT256_MINUS_HALF_PERCENTAGE / (PERCENTAGE_FACTOR - percentage))
+            PERCENTAGE_FACTOR - percentage == 0 ||
+                x <= (MAX_UINT256 - y * percentage - HALF_PERCENTAGE_FACTOR) / (PERCENTAGE_FACTOR - percentage)
```
Comment (QGarchery) [inline test/TestPercentageMath.sol:174]: The following is a general comment, and something I've been thinking about for the past weeks. So don't take it too litterally.

Are the test conditions really useful ? I mean if we rewrite them everytime we change the function, can we really say that it helped us maintain a certain behavior ? It is making development more difficult though, because we have to rewrite this test. I'm leaning more and more towards test that are more focused and durable, leaving behind completeness, as tests cannot be complete anyway. 

The general reasoning is that if we can write code more easily, then we can also fix the code more easily. Imagine a scenario for the same PR, but where you had to rewrite 100 tests like these (it would probably take more than a week of work in total). In this case, would you think it would be worth it to change this function at all ?

For this specific test, maybe we can restrict it to arbitrary large values like $10^{36}$

File: test/TestPercentageMath.sol [hunk 7]
```diff
@@ -201,6 +204,17 @@ contract TestPercentageMath is Test {
...
+    function testWeightedAvgOutOfBounds() public {
+        uint256 x = 5000;
+        uint256 y = 5000;
+        uint256 percentage = 1;
+
+        uint256 weightedAvg = PercentageMath.weightedAvg(x, y, percentage);
+
+        assertLe(x, weightedAvg);
+        assertLe(weightedAvg, y);
+    }
```
Comment (QGarchery) [inline test/TestPercentageMath.sol:216]: As an example, the idea of this test sounds really good on the other hand. Once we fixed the behavior in this PR we want to maintain it forever (or at least it should cost a bit to work around it)

**Review by QGarchery — COMMENTED** [review]
Nice ! The first condition allows you to check the second which is really cool. I think it does not even break symmetry: a call to `weightedAvg(a, b p)` reverts iff `weightedAvg(b, a, PERCENTAGE_FACTOR - p)` reverts. Not sure how to check that in solidity but it can be done in certora quite easily

File: test/TestPercentageMath.sol (hunk 6 above)
Comment (MathisGD) [reply test/TestPercentageMath.sol:174]: I asked me the same question.. this make few sense, because as you said I'm just re-writing the exact same conditions than in the code itself.

Comment (MathisGD) [reply test/TestPercentageMath.sol:174]: I'm not going to change everything in this PR, but let's open an issue for this, and fix it later on.

File: test/TestPercentageMath.sol (hunk 7 above)
Comment (MathisGD) [reply test/TestPercentageMath.sol:216]: yes I agree

File: test/TestPercentageMath.sol (hunk 6 above)
Comment (Rubilmax) [reply test/TestPercentageMath.sol:174]: Fair point ; I think it's a consequence of the fact that this test is a unit test whereas integration tests (those you describe) are much more generic
I agree it's taking more time to change the implementation of a function tested with unit tests (because of the issue you raised) ; though I think you're forgetting that these very tests are useful during development because they can (not always) shed light on changes and induced edge cases and forces the developer to double-check with unit testing
Most of the time, it's just double-checking
Sometimes (at least for me), it showed me some edge-cases

File: src/math/PercentageMath.sol (hunk 4 above)
Comment (MathisGD) [reply src/math/PercentageMath.sol:126]: oh yes great idea

File: test/TestPercentageMath.sol (hunk 6 above)
Comment (Rubilmax) [reply test/TestPercentageMath.sol:174]: These unit tests are not testing that the overall function works as expected, rather than each branch of the code are working as expected
I agree it's lacking an integration test which tests a subset of values, just as you described

Comment (MathisGD) [reply test/TestPercentageMath.sol:174]: >These unit tests are not testing that the overall function works as expected, rather than each branch of the code are working as expected

yes, but we mostly copy the conditions written in the code to separate these branches, so it is easy to do something wrong

Comment (QGarchery) [reply test/TestPercentageMath.sol:174]: Good points, let's discuss on the dedicated issue

File: src/math/PercentageMath.sol [hunk 8]
```diff
@@ -112,29 +112,23 @@ library PercentageMath {
...
-        //     x2 = y * percentage + HALF_PERCENTAGE_FACTOR > type(uint256).max
-        //     <=> y * percentage > type(uint256).max - HALF_PERCENTAGE_FACTOR
+        //     y * percentage + HALF_PERCENTAGE_FACTOR > type(uint256).max
         //     <=> percentage > 0 and y > (type(uint256).max - HALF_PERCENTAGE_FACTOR) / percentage
-        //
-        // Note: x1 / PERCENTAGE_FACTOR + x2 / PERCENTAGE_FACTOR <= x * (1 - p) + 1 + y * p + 1 where p is the fractional number percentage / PERCENTAGE_FACTOR
-        // <= max(x, y) + 2 <= type(uint256).max - HALF_PERCENTAGE_FACTOR + 2 <= type(uint256).max
+        // or if
+        //     x * (PERCENTAGE_FACTOR - percentage) + y * percentage + HALF_PERCENTAGE_FACTOR > type(uint256).max
+        //     <=> (PERCENTAGE_FACTOR - percentage) and (x > type(uint256).max - HALF_PERCENTAGE_FACTOR - y * percentage ) / (PERCENTAGE_FACTOR - percentage)
```
Comment (QGarchery) [inline src/math/PercentageMath.sol:119]:
```suggestion
        //     <=> (PERCENTAGE_FACTOR - percentage) > 0 and x > (type(uint256).max - HALF_PERCENTAGE_FACTOR - y * percentage) / (PERCENTAGE_FACTOR - percentage)
```

File: src/math/PercentageMath.sol [hunk 9]
```diff
@@ -112,29 +112,23 @@ library PercentageMath {
...
         assembly {
             z := sub(PERCENTAGE_FACTOR, percentage) // Temporary assignment to save gas.
-
             if or(
-                or(gt(percentage, PERCENTAGE_FACTOR), mul(z, gt(x, div(MAX_UINT256_MINUS_HALF_PERCENTAGE, z)))),
-                mul(percentage, gt(y, div(MAX_UINT256_MINUS_HALF_PERCENTAGE, percentage)))
+                gt(percentage, PERCENTAGE_FACTOR),
+                or(
+                    mul(percentage, gt(y, div(MAX_UINT256_MINUS_HALF_PERCENTAGE, percentage))),
+                    mul(z, gt(x, div(sub(sub(MAX_UINT256, HALF_PERCENTAGE_FACTOR), mul(y, percentage)), z)))
```
Comment (QGarchery) [inline src/math/PercentageMath.sol:126]: Maybe chose the same thing for the 2 lines, either `MAX_UINT256_MINUS_HALF_PERCENTAGE` or `sub(MAX_UINT256, HALF_PERCENTAGE_FACTOR)`

**Review by Rubilmax — APPROVED** [review]

**Review by QGarchery — APPROVED** [review]
LGTM ! I think it is correct now, meaning that it respects its specification given by "Executes a weighted average (x * (1 - p) + y * p), rounded up". As a consequence, it should be bound up by `max(x, y)` and bounded down by `min(x, y)`
